onnx/defs/optional/defs.cc

```diff
@@ -0,0 +1,139 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "onnx/defs/function.h"
+#include "onnx/defs/schema.h"
+
+#include <algorithm>
+#include <numeric>
+
+namespace ONNX_NAMESPACE {
+
+static const char* Optional_ver1_doc = R"DOC(
+Construct an optional type containing either an empty optional of a certain type specified by the attribute,
+or an optional type containing the 'input' element.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    Optional,
+    15,
+    OpSchema()
+        .SetDoc(Optional_ver1_doc)
+        .Input(0, "input", "The input element.", "V", OpSchema::Optional)
+        .Attr("type", "Type of the element in the optional output", AttributeProto::TYPE_PROTO, OPTIONAL_VALUE)
+        .Output(0, "output", "The optional output enclosing the input element.", "O")
+        .TypeConstraint(
+            "V",
+            [](){
+              auto t = OpSchema::all_tensor_types();
+              auto s = OpSchema::all_tensor_sequence_types();
+              t.insert(t.end(), s.begin(), s.end());
+              return t;
+            }(),
+            "Constrains input type to all tensor and sequence types.")
```

Comment:
General comment: We are forced to enumerate the set of allowed types here by ONNX's current design. This makes these ops more restrictive than necessary. We should generalize the implementation to allow unconstrained types as discussed in https://github.com/onnx/onnx/issues/2324. However, I doubt if that can be done by the deadline for current release, so I feel it is okay to have this in this PR for the short term.

Reply:
Makes sense. Thanks.